velox/tool/trace/PartitionedOutputReplayer.cpp

```diff
@@ -0,0 +1,167 @@
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/tool/trace/PartitionedOutputReplayer.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+
+namespace facebook::velox::tool::trace {
+namespace {
+std::shared_ptr<core::QueryCtx> createQueryContext(
+    const std::unordered_map<std::string, std::string>& config,
+    folly::Executor* executor) {
+  return core::QueryCtx::create(executor, core::QueryConfig(std::move(config)));
+}
+
+std::function<core::PlanNodePtr(std::string, core::PlanNodePtr)>
+partitionedOutputNodeFactory(const core::PartitionedOutputNode* originalNode) {
+  return [=](const core::PlanNodeId& nodeId,
+             const core::PlanNodePtr& source) -> core::PlanNodePtr {
+    return std::make_shared<core::PartitionedOutputNode>(
+        nodeId,
+        originalNode->kind(),
+        originalNode->keys(),
+        originalNode->numPartitions(),
+        originalNode->isReplicateNullsAndAny(),
+        originalNode->partitionFunctionSpecPtr(),
+        originalNode->outputType(),
+        source);
+  };
+}
+
+std::vector<std::unique_ptr<folly::IOBuf>> getData(
+    const std::shared_ptr<exec::OutputBufferManager>& bufferManager,
+    const std::string& taskId,
+    int destination,
+    int64_t sequence,
+    folly::Executor* executor) {
+  auto [promise, semiFuture] =
+      folly::makePromiseContract<std::vector<std::unique_ptr<folly::IOBuf>>>();
+  VELOX_CHECK(bufferManager->getData(
+      taskId,
+      destination,
+      exec::PartitionedOutput::kMinDestinationSize,
+      sequence,
+      [result = std::make_shared<
+           folly::Promise<std::vector<std::unique_ptr<folly::IOBuf>>>>(
+           std::move(promise))](
+          std::vector<std::unique_ptr<folly::IOBuf>> pages,
+          int64_t /*inSequence*/,
+          std::vector<int64_t> /*remainingBytes*/) {
+        result->setValue(std::move(pages));
+      }));
+  auto future = std::move(semiFuture).via(executor);
+  future.wait(std::chrono::seconds{10});
+  VELOX_CHECK(future.isReady());
+  return std::move(future).value();
+}
+} // namespace
+
+std::vector<std::vector<std::unique_ptr<folly::IOBuf>>> getAllData(
+    const std::shared_ptr<exec::OutputBufferManager>& bufferManager,
+    const std::string& taskId,
+    uint32_t numPartitions,
+    folly::Executor* executor) {
+  std::vector<std::thread> consumerThreads;
+  std::vector<std::vector<std::unique_ptr<folly::IOBuf>>> partitionedResults;
+  consumerThreads.reserve(numPartitions);
+  partitionedResults.reserve(numPartitions);
+  partitionedResults.resize(numPartitions);
+  for (uint32_t i = 0; i < numPartitions; i++) {
+    consumerThreads.push_back(std::thread([&, partition = i]() {
+      bool finished{false};
+      while (!finished) {
+        std::vector<std::unique_ptr<folly::IOBuf>> pages;
+        {
+          pages = getData(
```

Comment:
Not sure if we could hold this much data in memory when replay the trace from real production workload. Shall we just drop the received data or put the test value here if we want to capture the actual data. Or if we want to capture how many vectors or rows has been received. We could produce output row as metadata like table writer does.